lib/pacemaker/pcmk_cluster_queries.c

```diff
@@ -0,0 +1,366 @@
+               reply->data.ping.result);
+        exit_code = CRM_EX_OK;
+    }
+
+    event_done(controld_api);
+}
+
+static void
+designated_controller_event_cb(pcmk_ipc_api_t *controld_api,
+                    enum pcmk_ipc_event event_type, crm_exit_t status,
+                    void *event_data, void *user_data)
+{
+    pcmk__output_t *out = user_data;
+    pcmk_controld_api_reply_t *reply = controld_event_reply(out, controld_api,
+        event_type, status, event_data);
+
+    if (reply != NULL) {
+        out->message(out, "dc", reply->host_from);
+        exit_code = CRM_EX_OK;
+    }
+
+    event_done(controld_api);
+}
+
+static void
+pacemakerd_event_cb(pcmk_ipc_api_t *pacemakerd_api,
+                    enum pcmk_ipc_event event_type, crm_exit_t status,
+                    void *event_data, void *user_data)
+{
+    pcmk__output_t *out = user_data;
+    pcmk_pacemakerd_api_reply_t *reply = event_data;
+
+    crm_time_t *crm_when = crm_time_new(NULL);
+    char *pinged_buf = NULL;
+
+    switch (event_type) {
+        case pcmk_ipc_event_disconnect:
+            if (exit_code == CRM_EX_DISCONNECT) { // Unexpected
+                out->err(out, "error: Lost connection to pacemakerd");
+            }
+            event_done(pacemakerd_api);
+            return;
+
+        case pcmk_ipc_event_reply:
+            break;
+
+        default:
+            return;
+    }
+
+    if (message_timer_id != 0) {
+        g_source_remove(message_timer_id);
+        message_timer_id = 0;
+    }
+
+    if (status != CRM_EX_OK) {
+        out->err(out, "error: Bad reply from pacemakerd: %s",
+                crm_exit_str(status));
+        event_done(pacemakerd_api);
+        return;
+    }
+
+    if (reply->reply_type != pcmk_pacemakerd_reply_ping) {
+        out->err(out, "error: Unknown reply type %d from pacemakerd",
+                reply->reply_type);
+        event_done(pacemakerd_api);
+        return;
+    }
+
+    // Parse desired information from reply
+    crm_time_set_timet(crm_when, &reply->data.ping.last_good);
+    pinged_buf = crm_time_as_string(crm_when,
+        crm_time_log_date | crm_time_log_timeofday |
+            crm_time_log_with_timezone);
+
+    out->message(out, "pacemakerd-health",
+        reply->data.ping.sys_from,
+        (reply->data.ping.status == pcmk_rc_ok)?
+            pcmk_pacemakerd_api_daemon_state_enum2text(
+                reply->data.ping.state):"query failed",
+        (reply->data.ping.status == pcmk_rc_ok)?pinged_buf:"");
+    exit_code = CRM_EX_OK;
+    free(pinged_buf);
+
+    event_done(pacemakerd_api);
+}
+
+static pcmk_ipc_api_t *
+ipc_connect(pcmk__output_t *out, enum pcmk_ipc_server server, pcmk_ipc_callback_t cb)
+{
+    int rc;
+    pcmk_ipc_api_t *api = NULL;
+
+    rc = pcmk_new_ipc_api(&api, server);
+    if (api == NULL) {
+        out->err(out, "error: Could not connect to %s: %s",
+                (server == pcmk_ipc_controld) ? "controller" : "pacemakerd",
```

Comment:
We have a handy function for this: pcmk_ipc_name(api, true) (true = human-friendly name for logging, false = identifier used with libqb)